comparison_interface/configuration/validation.py

```diff
@@ -12,11 +17,69 @@ def __init__(self, app) -> None:
             self.__app.logger.critical(err)
             exit()
+        # now if we reference a csv file validate that
+        if "csvFile" in conf["comparisonConfiguration"]:
+            config_location = WS.get_configuration_location(self.__app)
+            # check the csv file structure is good enough.
+            self.validate_csv_structure(os.path.join(config_location, conf["comparisonConfiguration"]["csvFile"]))
+            self.__app.logger.info("structure of csv file is good")
+            # structure is fine so read the contents and send it to the comparisonConfiguration schema validator
+            config = CsvProcessor().create_config_from_csv(
+                os.path.join(config_location, conf["comparisonConfiguration"]["csvFile"])
+            )
+            schema = CompSchema()
+            try:
+                schema.load(config)
+            except ValidationError as err:
+                self.__app.logger.critical(err)
+                exit()
+
+    def check_config_path(self, path):
+        """Check that the path provided meets the requirements.
+
+        The requirements are either a path to a JSON file or a path to a directory containing a single JSON file and
+        a CSV file.
+        """
+        full_path = os.path.abspath(os.path.dirname(__file__)) + "/../" + path
+        if os.path.isdir(full_path):
+            file_count = 0
+            json_file = None
+            csv_file = None
+            for file in os.listdir(full_path):
+                file_count += 1
+                if file.lower()[-4:] == ".csv":
+                    csv_file = file
+                elif file.lower()[-5:] == ".json":
+                    json_file = file
+            if file_count != 2 or json_file is None or csv_file is None:
+                self.__app.logger.critical(
+                    "If the config path is to a directory then the directory must contain a JSON file and a CSV file."
+                )
+                exit()
+        elif os.path.isfile(full_path):
+            if path.lower()[-5:] != ".json":
+                self.__app.logger.critical("If the config path is to a file it must be a .json file.")
+                exit()
+
+    def validate_csv_structure(self, file):
+        """Validate the provided csv file."""
+        with open(file, mode='r') as csv_input:
+            image_data = DictReader(csv_input)
+            image_data.fieldnames = [x.lower() for x in image_data.fieldnames]
+            # required 'Item display name', 'image' everything else can be invented (case doesn't matter)
+            required_keys = ['item display name', 'image']
+            # optional_keys = ['item name', 'group name', 'group display name']
+            for key in required_keys:
+                if key not in image_data.fieldnames:
+                    raise ValidationError(
+                        'The csv file must have columns named "item display  name" and "image" (case does not matter, '
```

Comment:
Any reason why you use apostrophes (`'`) as the delimiter and quotes (`"`) for eg `"name"`? I understand that you can do this either way around and PEP8 doesn't dictate a preference, but also note that you do this the other way around eg line 156 in `schema.py`.
It doesn't bother me at all, but I was wondering if it might bother you? Just thinking about some of our conversations about being precise 😸

Reply:
It does annoy me, I will change it and any others I catch like the mode='r' just above it. The reason is I always use single quotes because of the amount of html and xml I write (which uses the double quote) but generally in Python people use double for strings and the code I inherited for this mostly uses double so I'm trying to conform and regularly revert to my usual style.